libs/language-server/src/lib/ast/wrappers/wrapper-factory-provider.ts

```diff
@@ -292,3 +315,93 @@ class TypedObjectWrapperFactory {
     assertUnreachable(type);
   }
 }
+
+class ValueTypeWrapperFactory {
+  constructor(
+    private readonly wrapperFactories: WrapperFactoryProvider,
+    private readonly primitiveValueTypeProvider: PrimitiveValueTypeProvider,
+  ) {}
+
+  wrap(
+    identifier: ValuetypeDefinition | ValueTypeReference | undefined,
+  ): ValueType | undefined {
+    if (identifier === undefined) {
+      return undefined;
+    } else if (isValueTypeReference(identifier)) {
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      const valueTypeDefinition = identifier?.reference?.ref;
+      if (valueTypeDefinition?.name === 'Collection') {
+        return this.wrapCollection(identifier);
+      }
+      return this.wrap(valueTypeDefinition);
+    } else if (isValuetypeDefinition(identifier)) {
+      if (identifier.name === 'Collection') {
+        // We don't have an object representing a generic collection
+        return;
+      }
+      if (identifier.isBuiltin) {
+        return this.wrapPrimitive(identifier);
+      }
+      return new AtomicValueType(
+        identifier,
+        this.primitiveValueTypeProvider,
+        this.wrapperFactories,
+      );
+    }
+    assertUnreachable(identifier);
+  }
+
+  wrapCollection(collectionRef: ValueTypeReference): CollectionValueType {
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    const collectionDefinition = collectionRef?.reference?.ref;
+    assert(collectionDefinition?.name === 'Collection');
+    const collectionGenerics = collectionRef.genericRefs;
+    if (collectionGenerics.length !== 1) {
+      throw new Error(
+        "Valuetype Collection needs exactly one generic parameter to define its elements' type",
+      );
+    }
+    const generic = collectionGenerics[0];
+    assert(generic !== undefined);
+    const elementValuetype = this.wrap(generic.ref);
+    if (elementValuetype === undefined) {
+      throw new Error(
+        "Could not create value type for the elements' type of value type Collection",
+      );
+    }
+    return new CollectionValueType(elementValuetype);
+  }
+
+  createCollection<I extends InternalValueRepresentation>(
+    input: ValueType<I>,
+  ): CollectionValueType<I> {
+    return new CollectionValueType(input);
+  }
+
+  wrapPrimitive(
+    builtinValuetype: ValuetypeDefinition,
+  ): PrimitiveValueType | undefined {
```

Comment:
To be discussed: naming okay so? Or should be "namespaced", e.g., `Valuetypes.Primitives.wrap` instead pf `Valuetypes.wrapPrimitives`.

Reply:
No I think this is fine.